proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,531 @@
+  allowed because the basic auth specification disallows them in usernames.  Kubernetes custom resource name validation
+  already enforces that this field must be a DNS subdomain, which means it must consist of lower case alphanumeric
+  characters, '-' or '.', and must start and end with an alphanumeric character (i.e. we do not need to do anything
+  special to enforce that clients do not have a ":" in their name).  See the audience confusion discussion below for
+  details on further restrictions that are applied to this field (i.e. required prefix).
+- `metadata.namespace`: Only clients in the same namespace as the Supervisor will be honored. This prevents cluster
+  users who have write permission in other namespaces from changing the configuration of the Supervisor.
+- `allowedRedirectURIs`: The list of allowed redirect URI. Must be `https://` URIs.  `127.0.0.1`, `localhost`, and other
+  forms of loopback redirect URIs are disallowed.
+- `allowedGrantTypes`: May only contain the following valid options:
+    - `authorization_code` allows the client to perform the authorization code grant flow, i.e. allows the webapp to
+      authenticate users.  This grant must always be listed.
+    - `refresh_token` allows the client to perform refresh grants for the user to extend the user's session.  This grant
+      must be listed if `allowedScopes` lists `offline_access`.
+    - `urn:ietf:params:oauth:grant-type:token-exchange` allows the client to perform RFC8693 token exchange, which is a
+      step in the process to be able to get a cluster credential for the user.  This grant must be listed if
+      `allowedScopes` lists `pinniped:request-audience`.
+- `allowedScopes`: Decide what the client is allowed to request. Note that the client must also actually request
+  particular scopes during the authorization flow for the scopes to be granted. May only contain the following valid
+  options:
+    - `openid`: The client is allowed to request ID tokens.  ID tokens only include the
+      [required claims](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) by default (`iss`, `sub`, `aud`,
+      `exp`, `iat`).  This scope must always be listed.
+    - `offline_access`: The client is allowed to request an initial refresh token during the authorization code grant
+      flow.  This scope must be listed if `allowedGrantTypes` lists `refresh_token`.
+    - `pinniped:request-audience`: The client is allowed to request a new audience value during a RFC8693 token
+      exchange, which is a step in the process to be able to get a cluster credential for the user.  `openid`,
+      `username` and `groups` scopes must be listed when this scope is present.  This scope must be listed if
+      `allowedGrantTypes` lists `urn:ietf:params:oauth:grant-type:token-exchange`.
+    - `username`: The client is allowed to request that ID tokens contain the user's username.  This is a newly
```

Comment:
I'm not sure that the `username` scope is necessary. I think it's acceptable to always include the username.

Reply:
> I'm not sure that the `username` scope is necessary. I think it's acceptable to always include the username.

The same reasoning could be applied to groups.  I think it is better to be specific about exactly what information is shared and when.  For example, in the future if we supported an `email` claim, a web app could be granted access to that claim but not the username.